sql: improve string formatting for postgres-compat

justinj opened the pull request:
Three main changes here:
* When formatting strings for arrays, postgres uses double quotes rather
  than single quotes.
* When outputting strings in bare-mode, postgres quotes both `NULL` and
  ``.
* When outputting strings within arrays, don't prefix strings with
  escape sequences with `e` (doesn't completely deal with #16487,
  though, as we are overly aggressive with double quotes, I think)

These three changes are required for the node.js postgres driver.

Commit:
sql: improve string formatting for postgres-compat

This change introduces a new method of formatting strings within arrays
which is more compatible with how Postgres formats them. It's also a
slightly simplified version of what Postgres does. The primary
difference is that we quote every string, and we use double quotes
instead of single quotes.

These changes are required for the node.js postgres driver.

I added tests to the python and node acceptance tests.
The ruby pg gem doesn't seem to support arrays at all, and adding a java
one is blocked on #17140.

I ran a quick benchmark comparing the two functions:

```
func BenchmarkEncodeSQLString(b *testing.B) {
	str := strings.Repeat("foo", 10000)
	b.Run("old version", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			encodeSQLStringWithFlags(bytes.NewBuffer(nil), str, FmtBareStrings)
		}
	})
	b.Run("new version", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			encodeSQLStringInsideArray(bytes.NewBuffer(nil), str)
		}
	})
}
```

and the results are that the new version is substantially faster, though
this isn't entirely fair since there are no escaped characters in the
benchmark, so the old version does a lot of unnecessary work. It is
reassuring that there isn't much, if anything, lost due to inserting
into the buffer char-by-char using WriteByte instead of using
WriteString.:

BenchmarkEncodeSQLString/old_version-8  2000 1035068 ns/op
BenchmarkEncodeSQLString/new_version-8  5000 304645 ns/op

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17069)
<!-- Reviewable:end -->

knz commented on pkg/sql/parser/encode.go:
```diff
@@ -106,6 +126,9 @@ func encodeSQLStringWithFlags(buf *bytes.Buffer, in string, f FmtFlags) {
 			if encodedChar := encodeMap[ch]; encodedChar != dontEscape {
```
maybe have the quoteChars in the map to collapse the two branches?

knz commented on pkg/sql/parser/encode.go:
```diff
@@ -170,6 +193,11 @@ func encodeSQLBytes(buf *bytes.Buffer, in string) {
 			buf.WriteByte('\\')
 			buf.WriteByte(encodedChar)
 			start = i + 1
+		} else if ch == '\'' {
```
ditto

knz commented on pkg/sql/parser/encode.go:
```diff
@@ -70,21 +77,34 @@ func encodeSQLStringWithFlags(buf *bytes.Buffer, in string, f FmtFlags) {
+	if f.doubleQuoteStrings {
+		quoteChar = '"'
+	}
+	if mustQuoteStringsMap[strings.ToUpper(in)] {
```
That's no good. 

1) I do not understand the special casing of the "NULL" string. It seems like this is the wrong place to check this.
2) You're mandating an allocation an scanning a potentially very large string just for two values that take 0 and 4 comparisons to check. Better have all 16 variants of "NULL" in the map and compare with `len(in) <= 4 && mqsm[in]`.

knz reviewed (commented):
The way the encodeSQLstring was modified is waaaaay too ad-hoc.
There is a fundamental problem I think with the handle of "NULL"

but beyond this I strongly recommend you just write a separate function.

Also please report before/after perf results on a mini-benchmark that calls encodeSQLstring many times and/or large strings.

jordanlewis commented:
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/encode.go, line 84 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17069#-KpHS9amB4yuDiGJatXw-r1-84:-KpHV7t3UuYghkEjIg2q:bfh63po) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2af831cc340dddb523c45d08c0a5dadb854a507/pkg/sql/parser/encode.go#L84)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

That's no good. 

1) I do not understand the special casing of the "NULL" string. It seems like this is the wrong place to check this.
2) You're mandating an allocation an scanning a potentially very large string just for two values that take 0 and 4 comparisons to check. Better have all 16 variants of "NULL" in the map and compare with `len(in) <= 4 && mqsm[in]`.


</blockquote></details>

The special casing is important because Postgres double quotes any string whose value is a string named `null`, but agreed that this seems less efficient than it could be.

---

*[pkg/sql/parser/encode.go, line 129 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17069#-KpHTSn6O13b1a4Mia3E:-KpHTSn6O13b1a4Mia3F:b-ybzagw) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2af831cc340dddb523c45d08c0a5dadb854a507/pkg/sql/parser/encode.go#L129)):*
> ```Go
> 				buf.WriteByte('\\')
> 				buf.WriteByte(encodedChar)
> 			} else if ch == quoteChar {
> ```

Hmm does this change current behavior? If quoteChar is `'`, would we print `e'foo\''` as `foo\'` instead of `foo'`?

---

*[pkg/sql/pgwire/types.go, line 201 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17069#-KpHUQGjvVCBIEyRShNM:-KpHUQGjvVCBIEyRShNN:b-rw7rzj) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2af831cc340dddb523c45d08c0a5dadb854a507/pkg/sql/pgwire/types.go#L201)):*
> ```Go
> 			// TODO(radu): we are relying on Format but this doesn't work correctly
> 			// if we have an array inside this array. To support nested arrays, we
> 			// would need to recurse or add a special FmtFlag.
> ```

Is this todo still necessary?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17069)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/encode.go, line 84 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17069#-KpHS9amB4yuDiGJatXw-r1-84:-KpHWlnptAiRjNz-_d40:b6syq3b) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2af831cc340dddb523c45d08c0a5dadb854a507/pkg/sql/parser/encode.go#L84)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

The special casing is important because Postgres double quotes any string whose value is a string named `null`, but agreed that this seems less efficient than it could be.
</blockquote></details>

That's not true:
```
kena@jepsen-master ~/cockroach % psql
psql (9.5.7)
Type "help" for help.

kena=# select 'null';
 ?column?
----------
 null
(1 row)

```
I don't see any double quotes here.

This quoting only occurs in the context of string arrays. Therefore I think the function that pretty-prints the array should check the individual string elements and add the quoting when needed - not here in encodeSQLstring,

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17069)*
<!-- Sent from Reviewable.io -->

justinj commented:
Agreed with all these comments. I was hoping to avoid writing a separate function but I guess although the existing logic is somewhat complex, the behaviour here is different enough that trying to reuse too much of this is probably not worth it.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/encode.go, line 84 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17069#-KpHS9amB4yuDiGJatXw-r1-84:-KpHziPm3rBuVLpulkXe:bkzqaow) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2af831cc340dddb523c45d08c0a5dadb854a507/pkg/sql/parser/encode.go#L84)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

That's not true:
```
kena@jepsen-master ~/cockroach % psql
psql (9.5.7)
Type "help" for help.

kena=# select 'null';
 ?column?
----------
 null
(1 row)

```
I don't see any double quotes here.

This quoting only occurs in the context of string arrays. Therefore I think the function that pretty-prints the array should check the individual string elements and add the quoting when needed - not here in encodeSQLstring,
</blockquote></details>

Good point about the behaviour, I got mixed up between the within-array and not-within-array differences :cry: and agreed on the upcasing, I guess just including all variations is the way to go.

---

*[pkg/sql/parser/encode.go, line 129 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17069#-KpHTSn6O13b1a4Mia3E:-KpI-Nj_GHXyHoDCxDR9:b-nxk2dj) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2af831cc340dddb523c45d08c0a5dadb854a507/pkg/sql/parser/encode.go#L129)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Hmm does this change current behavior? If quoteChar is `'`, would we print `e'foo\''` as `foo\'` instead of `foo'`?
</blockquote></details>

Yes, I think you're right that this is wrong, I'll be sure to add a test for this case.

As an aside it seems that `psql`'s (the cli, not postgres itself) handling of `select array['foo\''];` seems broken unless I'm misunderstanding something? It just breaks to a new line?

---

*[pkg/sql/pgwire/types.go, line 201 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17069#-KpHUQGjvVCBIEyRShNM:-KpI-k0Hs0NIHMvNaHTI:b-2ghyk8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2af831cc340dddb523c45d08c0a5dadb854a507/pkg/sql/pgwire/types.go#L201)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Is this todo still necessary?
</blockquote></details>

I believe so, unfortunately. The `FormatNode` call below formats the values for an AST, so (though I haven't tested) it would format any arrays as `ARRAY[...]` rather than `{...}`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17069#-:-KpI00HwE8InLCeI-whi:bvhbtj8)*
<!-- Sent from Reviewable.io -->

justinj commented:
PTAL :/ unhappy about the massive duplication of this pretty complex function but I can't come up with anything better. It's not *too* bad I guess, since this code is somewhat general and not really dependent on anything.

Still doesn't fix 16487, that is indeed an annoying one.

---

Review status: 1 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/sql/parser/encode.go, line 129 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17069#-KpHTSn6O13b1a4Mia3E:-KpLbwdY4N08RAPeOxs0:bvvrshv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2af831cc340dddb523c45d08c0a5dadb854a507/pkg/sql/parser/encode.go#L129)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Yes, I think you're right that this is wrong, I'll be sure to add a test for this case.

As an aside it seems that `psql`'s (the cli, not postgres itself) handling of `select array['foo\''];` seems broken unless I'm misunderstanding something? It just breaks to a new line?
</blockquote></details>

follow up - I was wrong, you need to prefix with the `e` sigil to get pg to accept that input.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17069#-:-KpLty03OTMY7N_W17KL:bvmw2nf)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 1 of 5 files at r1, 3 of 4 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/sql/parser/encode.go, line 165 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17069#-KpLx3c9GHAPpBMu3DpV:-KpLx3cA-svRgXTebeNa:b-kupoxx) ([raw file](https://github.com/cockroachdb/cockroach/blob/5dc692039dfb4273df2db6f1e786186ea6f8a9e8/pkg/sql/parser/encode.go#L165)):*
> ```Go
> 	}
> 	in := string(MustBeDString(d))
> 	if mustQuoteStringsMap[in] {
> ```

I still don't think this is the right way to do this.
First of all I am not convinced you need to special case strings containing the word "null". You'd need an extensive comment in the code and the commit message to motivate that.

Then if you *really* do then this makes much more sense to me:
```go
if len(s) == 0 || (len(s) == 4 && 
    (s[0] == 'n' || s[0] == 'N') &&
    (s[1] == 'u' || s[0] == 'U') &&
    (s[2] == 'l' || s[0] == 'L') &&
    (s[3] == 'l' || s[0] == 'L')) { ... }
```

---

*[pkg/sql/pgwire/types.go, line 201 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17069#-KpHUQGjvVCBIEyRShNM:-KpLwYJCWCi5azxrSTg7:b-2ivqrs) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2af831cc340dddb523c45d08c0a5dadb854a507/pkg/sql/pgwire/types.go#L201)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I believe so, unfortunately. The `FormatNode` call below formats the values for an AST, so (though I haven't tested) it would format any arrays as `ARRAY[...]` rather than `{...}`.
</blockquote></details>

What Radu says is what you should be doing here instead of this `UnwrapType` business above.
Make a format flag to choose which way string literals are printed. Then use this flag in the recursion.
Let me know if you need help with achieving this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17069)*
<!-- Sent from Reviewable.io -->

justinj commented:
I'm going to have to add a bunch of integration tests to go along with this change, but PTAL in the meantime! Didn't include recursing to support multidimensional arrays since there's no good way to test it yet.

From the (new) commit message:

I ran a quick benchmark comparing the two functions:

```
func BenchmarkEncodeSQLString(b *testing.B) {
	str := strings.Repeat("foo", 10000)
	b.Run("old version", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			encodeSQLStringWithFlags(bytes.NewBuffer(nil), str, FmtBareStrings)
		}
	})
	b.Run("new version", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			encodeSQLStringInsideArray(bytes.NewBuffer(nil), str)
		}
	})
}
```

and the results are that the new version is substantially faster, though
this isn't entirely fair since there are no escaped characters in the
benchmark, so the old version does a lot of unnecessary work. It is
reassuring that there isn't much, if anything, lost due to inserting
into the buffer char-by-char using WriteByte instead of using
WriteString:

BenchmarkEncodeSQLString/old_version-8  2000 1035068 ns/op
BenchmarkEncodeSQLString/new_version-8  5000 304645 ns/op

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with nits

---

Reviewed 18 of 18 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/encode.go, line 148 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17069#-KpS3E2ISw6Pqww_8Wz2:-KpS3E2JsmuUQWzOFP11:b-ty9h21) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c1e79599d2a2d4c65f6bc6dbbf4d4f8a129f138/pkg/sql/parser/encode.go#L148)):*
> ```Go
> }
> 
> // EncodeSQLStringInsideArray writes a string literal to buf using the "string
> ```

lowercase identifier in comment

---

*[pkg/sql/pgwire/types.go, line 199 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17069#-KpS3V76AVfNATYnrQhx:-KpS3V76AVfNATYnrQhy:b-8wlfpd) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c1e79599d2a2d4c65f6bc6dbbf4d4f8a129f138/pkg/sql/pgwire/types.go#L199)):*
> ```Go
> 				b.variablePutbuf.WriteString(sep)
> 			}
> 			// TODO(radu): we are relying on Format but this doesn't work correctly
> ```

The TODO can go. Or perhaps transform this to `TODO(justin): add a test for nested arrays.`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17069#-:-KpS3ge4wjqXzedXau7A:bfmeux1)*
<!-- Sent from Reviewable.io -->

justinj commented:
Ok, fixed the nits and added integration tests where possible, going to merge when green.

---

Review status: 8 of 20 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/sql/pgwire/types.go, line 199 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17069#-KpS3V76AVfNATYnrQhx:-KpV6ZjMGRk8IMba9NeM:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c1e79599d2a2d4c65f6bc6dbbf4d4f8a129f138/pkg/sql/pgwire/types.go#L199)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

The TODO can go. Or perhaps transform this to `TODO(justin): add a test for nested arrays.`
</blockquote></details>

:thumbsup: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17069#-:-KpW1fRih_zd4F_ZphtZ:b-h28uku)*
<!-- Sent from Reviewable.io -->
